internal/destination/destination.go

```diff
@@ -131,29 +135,44 @@ func (d *Destination) Open(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("connect to NATS: %w", err)
 	}
-
-	d.writer, err = NewWriter(WriterParams{
-		Conn:          conn,
-		Subject:       d.config.Subject,
-		RetryWait:     d.config.RetryWait,
-		RetryAttempts: d.config.RetryAttempts,
-	})
-	if err != nil {
-		return fmt.Errorf("init jetstream writer: %w", err)
-	}
+	d.nc = conn
 
 	// Async handlers & callbacks
 	conn.SetErrorHandler(internal.ErrorHandlerCallback(ctx))
-	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {}))
-	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {}))
+	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {
+		d.writer.stopWrites()
+	}))
+	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {
+		d.writer, err = NewWriter(writerParams{
+			nc:            d.nc,
+			subject:       d.config.Subject,
+			retryWait:     d.config.RetryWait,
+			retryAttempts: d.config.RetryAttempts,
+		})
+		d.writer.startWrites()
+	}))
 	conn.SetClosedHandler(internal.ClosedCallback(ctx))
 	conn.SetDiscoveredServersHandler(internal.DiscoveredServersCallback(ctx))
 
+	d.writer, err = NewWriter(writerParams{
+		nc:            d.nc,
+		subject:       d.config.Subject,
+		retryWait:     d.config.RetryWait,
+		retryAttempts: d.config.RetryAttempts,
+	})
+	if err != nil {
+		return fmt.Errorf("init jetstream writer: %w", err)
+	}
+
 	return nil
 }
 
 // Write writes a record into a Destination.
 func (d *Destination) Write(_ context.Context, records []sdk.Record) (int, error) {
+	if !d.nc.IsConnected() {
+		return -1, errWriteUnavailable
```

Comment:
Returning an error from `Write` will cause the pipeline to stop, I don't think this is the behavior we're aiming for, right? IIUC we rather want to wait until we reconnect (or the context gets cancelled) and write the records at that point.

Also, the returned integer represents the number of successfully written records, that's how Conduit knows how many records to ack / nack, if no records were written it should return 0.

Reply:
Ok, changing to Zero.

Question: if a write fails, does Conduit retries that?

Reply:
No, Conduit is pretty strict this way, if it gets an error it stops the pipeline. The good news is the connector can block as long as it wants in `Write` as long it respects the context. So if the context gets cancelled while it's waiting for the reconnection, then it should return an error, which will tell Conduit that the messages didn't get written and they will be nacked.